src/app/components/hooks/use-model-state.ts

```diff
@@ -87,76 +89,127 @@ export const useModelState = <IModelInputState, IModelOutputState>(
       return newState;
     });
-  }, [activeRunIdx]);
-
+  }, [activeRunIdx, saveInteractiveState]);
 
   const setOutputState = useCallback((update: Partial<IModelOutputState>) => {
-    setCurrentOutputState(oldState => ({...oldState, ...update}));
-  }, []);
+    setCurrentOutputState(oldState => {
+      saveInteractiveState({outputState: {...oldState, ...update}});
+      return {...oldState, ...update};
+    });
+  }, [saveInteractiveState]);
 
   const snapshotOutputState = useCallback((outputState: IModelOutputState) => {
     setModelRuns(oldState => {
       const newState = [...oldState];
+
       newState[activeRunIdx] = {
         ...newState[activeRunIdx],
         outputStateSnapshots: [...newState[activeRunIdx].outputStateSnapshots, outputState]
       };
+
+      saveInteractiveState({
+        inputState: {...newState[activeRunIdx].inputState},
+        outputState,
+        modelRuns: newState
+      });
+
       return newState;
     });
-  }, [activeRunIdx]);
+  }, [activeRunIdx, saveInteractiveState]);
 
   const addModelRun = useCallback(() => {
     setCurrentOutputState(initialOutputStateObject);
     setActiveOutputSnapshotIdx(null);
-    setModelRuns(oldState => [...oldState, getNewModelRun()]);
+
+    setModelRuns(oldState => {
+      const newState = [...oldState, getNewModelRun()];
+      saveInteractiveState({
+        inputState: initialInputState,
+        outputState: initialOutputStateObject,
+        modelRuns: newState
+      });
+      return newState;
+      }
+    );
+
     setActiveRunIdx(modelRuns.length);
-  }, [getNewModelRun, initialOutputStateObject, modelRuns.length]);
+  }, [getNewModelRun, initialInputState, initialOutputStateObject, modelRuns.length, saveInteractiveState]);
 
   // This will remove all existing runs and and create a new one.
   const removeAllModelRuns = useCallback(() => {
-    setCurrentOutputState(initialOutputStateObject);
+    setCurrentOutputState(() => {
+      saveInteractiveState({outputState: initialOutputStateObject});
+      return initialOutputStateObject;
+    });
+
     setActiveRunIdx(0);
     setActiveOutputSnapshotIdx(null);
-    setModelRuns([getNewModelRun()]);
-  }, [getNewModelRun, initialOutputStateObject]);
+
+    setModelRuns(() => {
+      saveInteractiveState({modelRuns: [getNewModelRun()]});
+      return [getNewModelRun()];
+    });
+  }, [getNewModelRun, initialOutputStateObject, saveInteractiveState]);
 
   const removeModelRun = useCallback(() => {
     if (modelRuns.length > 1) {
       const newActiveRunIdx = activeRunIdx > 0 ? activeRunIdx - 1 : activeRunIdx + 1;
       const newActiveRun = modelRuns[newActiveRunIdx];
       const snapshots = newActiveRun.outputStateSnapshots;
-      setCurrentOutputState(snapshots[snapshots.length - 1]);
+
+      setCurrentOutputState(() => {
+        saveInteractiveState({outputState: snapshots[snapshots.length - 1]});
```

Comment:
If `snapshots.length` is 0 here, that will make `outputState` undefined but shouldn't it be the initial output state in that case?